### PR TITLE
e2e playwright test

### LINKED ISSUE #11: Add end to end testing
We want to test the final `nix build` web app. 

The official leptos templates use [playwright](https://playwright.dev/) to do this:

https://github.com/leptos-rs/start-axum/tree/main/end2end

While we may do the same, I'd like to first explore (briefly) the option of being able to write tests in pure Rust.  For eg., there is playwright-rust, but it seems [unmaintained](https://github.com/octaltree/playwright-rust/pull/49#issuecomment-1564119957). Are there other options? 

Exhausting all that, we may just use something like playwright.

### PR DESCRIPTION
**Pull request by divinenaman**
fixes #11 using playwright

- [x] Initial impl
- [x] Github Actions CI check
- [x] VSCode IDE support w/ instructions

### CODE REVIEW

File: src/server.rs
Comment (srid) [inline src/server.rs:1]: We shouldn't be modifying the API endpoints just for e2e test. The changes to `server.rs` should be reverted.

File: e2e-playwright/tests/nix-version.test.js
Comment (srid) [inline e2e-playwright/tests/nix-version.test.js:8]: We want to test that **some** version is shown in the browser, and not that the _correct_ version is displayed. Thus, there is no need to poll any API.

File: Cargo.lock
Comment (srid) [inline Cargo.lock:1]: Why does this PR have changes to `Cargo.lock`?

**Review by srid — COMMENTED** [review]
@divinenaman The PR is coming along well. I've left some comments. Could you add Github Actions workflow to run the e2e tests in CI?

File: e2e-playwright/flake-module.nix
Comment (srid) [inline e2e-playwright/flake-module.nix:6]:
```suggestion
      process-compose.e2e-playwright-test =
```

File: Cargo.lock
Comment (divinenaman) [reply Cargo.lock:1]: it was giving errors

**Comment by divinenaman** [comment]
@srid is there a way to mock, github actions ? I remember it took around 10 commits last time I did something similar

**Comment by srid** [comment]
> @srid is there a way to mock, github actions ? I remember it took around 10 commits last time I did something similar

I don't know, but you do have write access to this repo, so you can push this branch to this repo (rather than your fork).

File: justfile
Comment (srid) [inline justfile:20]: Why was `cargo watch -- cargo leptos test` replaced?

File: e2e-playwright/tests/nix-version.test.js
Comment (srid) [inline e2e-playwright/tests/nix-version.test.js:1]: @divinenaman How is the IDE experience in VSCode like when editing this file?

File: justfile
Comment (srid) [reply justfile:20]: Oh, I just realized that tests were broken on `main`. I just fixed it in 855e964924bf94ea78aad04b16add08512db7e34 - you want to merge that.

File: e2e-playwright/tests/nix-version.test.js
Comment (divinenaman) [reply e2e-playwright/tests/nix-version.test.js:1]: @srid  I am just happy with onSaveFormat, for me it's good enough

Comment (srid) [reply e2e-playwright/tests/nix-version.test.js:1]: If we introduced `package.json` and the like, would it provide an improved experience (like autocompletion, and hover tooltips)?

Comment (srid) [reply e2e-playwright/tests/nix-version.test.js:1]: Let's do that, and explain that in `e2e-playwright/README.md`. We want good IDE experience for writing tests as well.

File: Cargo.lock
Comment (srid) [reply Cargo.lock:1]: @divinenaman The changes to `Cargo.lock` should be reverted. If you are seeing any errors, try to resolve them by finding the root cause - or post it here for help.

File: src/server.rs
Comment (srid) [inline src/server.rs:12]: Should be reverted

File: Cargo.lock
Comment (srid) [reply Cargo.lock:1]: Note: you can do this using `git show main:Cargo.lock > Cargo.lock`

File: e2e-playwright/tests/nix-version.test.js
Comment (srid) [reply e2e-playwright/tests/nix-version.test.js:1]: @divinenaman How do I get IDE support in vscode? Can you explain the steps?

This is what I see:

image

File: Cargo.lock
Comment (divinenaman) [reply Cargo.lock:1]: @srid reverted the changes, working as expected

File: e2e-playwright/tests/nix-version.test.js
Comment (srid) [reply e2e-playwright/tests/nix-version.test.js:1]: I was able to fix this by running `npm i` as well as adding the `package-lock.json`: https://github.com/juspay/nix-browser/commit/021e936e2468e88b0a3165696de46d1d4a322e38

However, it broke `nix run`:

image

@divinenaman Can you figure out a solution here? This is the only thing to do before merging this PR.

Comment (divinenaman) [reply e2e-playwright/tests/nix-version.test.js:1]: @srid I am not able to re-create, I did `nix-store --gc` to clear my nix store but tests passed for me, any hints to recreate ?

Comment (divinenaman) [reply e2e-playwright/tests/nix-version.test.js:1]: @srid can you once try now, added playwright browser path

Comment (srid) [reply e2e-playwright/tests/nix-version.test.js:1]: > also remove the local node-modules

It works without local node-modules.  But this local node-modules (created `npm i`) exists solely so that VSCode IDE support will work. So how would you make both work at the same time?

Comment (srid) [reply e2e-playwright/tests/nix-version.test.js:1]: > You have two different versions of @playwright/test. This usually happens
> when one of the dependencies in your package.json depends on @playwright/test.

This seems to be the culprit.

Comment (srid) [reply e2e-playwright/tests/nix-version.test.js:1]: I think I have a solution, hold on.

**Review by srid — APPROVED** [review]
Awesome, everything's good!

**Pull request merged**
